src/pagination.rs

```diff
@@ -0,0 +1,312 @@
+            // `order.unwrap_or(SortOrder::Desc);` so it's unreachable
+            unreachable!("Sort order should have default value when constructing PaginationConfig")
+        }
+    };
+    posts.sort_by(|a, b| {
+        let keys = &config.sort_by;
+        let mut cmp = Ordering::Less;
+        for k in keys {
+            cmp = match (extract_value(a, &k), extract_value(b, &k)) {
+                (Some(a), Some(b)) => order(a, b),
+                (None, None) => Ordering::Equal,
+                (_, None) => Ordering::Greater,
+                (None, _) => Ordering::Less,
+            };
+            if cmp != Ordering::Equal {
+                return cmp;
+            }
+        }
+        cmp
+    });
+}
+
+fn create_all_paginators(
+    mut all_posts: &mut Vec<liquid::value::Value>,
+    doc: &Document,
+    pagination_cfg: &PaginationConfig,
+) -> Result<Vec<Paginator>> {
+    let total_pages = all_posts.len();
+    // f32 used here in order to not lose information to ceil the result,
+    // otherwise we can lose an index
+    let total_indexes = (total_pages as f32 / pagination_cfg.per_page as f32).ceil() as usize;
+    let mut paginators = Vec::with_capacity(total_indexes);
+    // keeping `for` loop instead of `map` as asked in PR because I need `i`
+    // anyway and it keep the code more understandable. The issue was
+    // pre-allocating  which is done with `with_capacity`
+    for i in 0..total_indexes {
+        paginators.push(create_paginator(
+            i,
+            total_indexes,
+            total_pages,
+            &pagination_cfg,
+            &doc,
+            &mut all_posts,
+        )?);
+    }
+    Ok(paginators)
+}
+
+fn create_paginator(
+    i: usize,
+    total_indexes: usize,
+    total_pages: usize,
+    config: &PaginationConfig,
+    doc: &Document,
+    mut all_posts: &mut Vec<liquid::value::Value>,
+) -> Result<Paginator> {
+    let index = i + 1;
+    let mut paginator = Paginator::new(total_indexes, total_pages);
+
+    paginator.set_first_last(&doc, &config, total_indexes)?;
+    paginator.set_current_index_info(index, &mut all_posts, &config, &doc)?;
+    paginator.set_previous_next_info(index, total_indexes, &doc, &config)?;
+
+    Ok(paginator)
+}
+
+fn pagination_attributes(index_num: i32, include: Include) -> liquid::value::Object {
+    let mut attributes = liquid::value::Object::new();
+    attributes.insert("num".into(), liquid::value::Value::scalar(index_num));
+    let i: &str = include.into();
+    attributes.insert("include".into(), liquid::value::Value::scalar(i));
+    attributes
+}
+
+fn interpret_permalink(
+    config: &PaginationConfig,
+    doc: &Document,
+    index_num: usize,
+) -> Result<String> {
+    Ok(if index_num == 1 {
+        doc.url_path.clone()
+    } else {
+        let mut attributes = document::permalink_attributes(&doc.front, &doc.file_path);
+        let pagination_attr = pagination_attributes(index_num as i32, config.include);
```

Comment:
*(Observation only)*

Liquid's new `Globals` trait would help a lot here.  We'd no longer need to be creating a liquid Object.  Once I switch it to returning `Cow`, it'll be even better because we can then lazily create the items we need since most people probably don't use most of the attributes we create.

Reply:
Liquid no longer accepts `Object`  but `Globals` (`Object` just happens to implement it).

We could have a struct that contains all the relevant root objects and implements `Globals::get`.  It then has to manually map names to struct members (I'd love to make a derive macro for this).